modules/gapi/src/compiler/passes/match_pattern_to_substitute.cpp

```diff
@@ -0,0 +1,106 @@
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+#include "pattern_matching.hpp"
+
+#include <ade/util/zip_range.hpp>
+
+namespace cv { namespace gimpl {
+
+namespace {
+using Graph = GModel::Graph;
+using Metadata = typename Graph::CMetadataT;
+using VisitedMatchings = std::list<std::pair<ade::NodeHandle, ade::NodeHandle>>;
+
+// Returns true if two DATA nodes are semantically identical:
+//  - both nodes have the same GShape
+//  - both nodes have the same storage
+//
+// @param firstMeta - metadata of first
+// @param secondMeta - metadata of second
+bool compareDataNodes(const Metadata& firstMeta,
+                      const Metadata& secondMeta)
+{
+    GAPI_Assert(firstMeta.get<NodeType>().t == NodeType::DATA);
+    GAPI_Assert(firstMeta.get<NodeType>().t == secondMeta.get<NodeType>().t);
+
+    const auto& firstData = firstMeta.get<Data>();
+    const auto& secondData = secondMeta.get<Data>();
+    // compare shape
+    if (firstData.shape != secondData.shape) {
+        return false;
+    }
+    // compare storage
+    if (firstData.storage != secondData.storage) {
+        return false;
+    }
+
+    // FIXME: is it enough to only check shape && storage?
+    return true;
+}
+
+// Returns matched pairs of {pattern node, substitute node}
+SubgraphMatch::M matchDataNodes(const Graph& pattern,
+                                const Graph& substitute,
+                                const std::vector<ade::NodeHandle>& patternNodes,
+                                std::vector<ade::NodeHandle> substituteNodes)
+{
+    SubgraphMatch::M matched;
+    for (auto it : ade::util::zip(patternNodes, substituteNodes)) {
+        const auto& pNode = std::get<0>(it);
+        const auto& sNode = std::get<1>(it);
+        if (!compareDataNodes(pattern.metadata(pNode), substitute.metadata(sNode))) {
+            return {};
+        }
+        matched.insert({ pNode, sNode });
+    }
+    return matched;
+}
+}  // anonymous namespace
+
+SubgraphMatch matchPatternToSubstitute(const cv::gimpl::GModel::Graph& pattern,
+                                       const cv::gimpl::GModel::Graph& substitute,
+                                       const cv::gimpl::Protocol& patternP,
+                                       const cv::gimpl::Protocol& substituteP)
+{
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
+        || patternDataOutputs.size() != substituteDataOutputs.size()) {
+        return {};
+    }
```

Comment:
I have an question what do we intend to check here: whether user created a bad substitution or not?
I suppose there shall be check only on unique data nodes, because  Protocol::in_nhs can contain the copies of one node if this node is consumed more than once.
Checking only unique input data nodes can solve the next problems:

Pattern                   Substitute
   +--+  +--+          +--+
   |      |  |      |          |       |
   +--+  +--+          +--+
     |          |                 | |
     v         v                vv
   XXXX   XXXX         XXXX
  XX  XX XX  XX       XX  XX
   XXXX   XXXX         XXXX
      +         +             + +->---+
      v          v              v       |     |
   +--+   +--+         +--+   +--+
   |      |   |      |         |      |
   +--+   +--+         +--+

These two will  be matched but have not to.


Pattern                     Substitute
          +--+                    +--+
 +----+    +----+            |     |
 |        +--+        |           +-++
 |                        |               |
 |                        |               v
 |        XXXX       |            XXXX
+--> XX  XX<--+          XX  XX
          XXXX                    XXXX
             +                          +
             v                           v
          +-++                    +-++
           |     |                      |     |
          +--+                     +--+

These have to be matched but will not.

Reply:
As I got Protocol::in_nhs may contain copies only if we pass copies to the cv::GIn as input API, could ypu confirm this?

Reply:
To me it seems like:
1st pair of graphs: won't be matched - diff. number of ins (which makes sense)
2nd pair of graphs: will be matched if shape/storage are the same

I think that Protocol::in_nhs relates to the number of nodes passed to `GIn()`, so if it's the same node - fine, there will be as many as there are `GIn`s